corehq/apps/locations/bulk_management.py

```diff
@@ -187,18 +189,42 @@ def _optional_attr(attr):
                    data_model, delete_uncategorized_data)
         return stub
 
-    def lookup_old_collection_data(self, old_collection):
-        # Lookup whether the location already exists in old_collection or is new.
-        #   Depending on that set attributes like 'is_new', 'needs_save', 'db_obect'
-        self.autoset_location_id_or_site_code(old_collection)
-
+    @property
+    def new_parent(self):
+        if self._new_parent_stub is None:
+            return None
+        assert self._new_parent_stub.db_object is not None, \
+            'lookup_old_collection_data not called yet?'
+        return self._new_parent_stub.db_object
+
+    def lookup_old_collection_data(self, old_collection, locs_by_code):
+        """Lookup whether the location already exists or is new"""
         if self.is_new:
             self.db_object = SQLLocation(domain=old_collection.domain_name)
         else:
             self.db_object = copy.copy(old_collection.locations_by_id[self.location_id])
 
-        self.needs_save = self._needs_save()
-        self.moved_to_root = self._moved_to_root()
+        if self.db_object.parent_id is not None:
+            parent = old_collection.locations_by_pk[self.db_object.parent_id]
+            old_parent_code = parent.site_code
+        else:
+            old_parent_code = ROOT_LOCATION_TYPE
+        self.needs_save = self._needs_save(old_parent_code)
+        self.moved_to_root = (
+            not self.is_new
+            and self.parent_code == ROOT_LOCATION_TYPE
+            and old_parent_code != ROOT_LOCATION_TYPE
+        )
+        if self.parent_code and self.parent_code != ROOT_LOCATION_TYPE:
+            try:
+                self._new_parent_stub = locs_by_code[self.parent_code]
+            except KeyError:
+                # TODO require all referenced locations to be in stubs and then
+                # remove this exception handler and fake stub nonsense.
+                # Breaks test: TestBulkManagement.test_large_upload
```

Comment:
Not planning to do this in this PR, but I'm surprised this is not already required.